File: packages/python-sdk/plinto/http_client.py
"""
HTTP client for the Plinto SDK with automatic token refresh and error handling
"""

import asyncio
import json
import time
from typing import Any, Dict, Optional, Union
from urllib.parse import urljoin

import httpx
from httpx import Response

from .exceptions import (
    AuthenticationError,
    NetworkError,
    NotFoundError,
    PermissionError,
    RateLimitError,
    ServerError,
    ValidationError,
    PlintoError,
)


class TokenStorage:
    """Simple in-memory token storage"""
    
    def __init__(self):
        self._access_token: Optional[str] = None
        self._refresh_token: Optional[str] = None
        self._expires_at: Optional[float] = None
    
    def set_tokens(
        self,
        access_token: str,
        refresh_token: str,
        expires_in: int,
    ) -> None:
        """Store tokens with expiration time"""
        self._access_token = access_token
        self._refresh_token = refresh_token
        self._expires_at = time.time() + expires_in - 60  # Refresh 1 minute early
    
    def get_access_token(self) -> Optional[str]:
        """Get access token if not expired"""
        if self._access_token and self._expires_at:
            if time.time() < self._expires_at:
                return self._access_token
        return None
    
    def get_refresh_token(self) -> Optional[str]:
        """Get refresh token"""
        return self._refresh_token
    
    def clear_tokens(self) -> None:
        """Clear all stored tokens"""
        self._access_token = None
        self._refresh_token = None
        self._expires_at = None
    
    def is_expired(self) -> bool:
        """Check if access token is expired"""
        if not self._expires_at:
            return True
        return time.time() >= self._expires_at


class HTTPClient:
    """HTTP client with authentication and error handling"""
    
    def __init__(
        self,
        base_url: str,
        api_key: Optional[str] = None,
        timeout: int = 30,
        max_retries: int = 3,
        retry_delay: float = 1.0,
    ):
        self.base_url = base_url.rstrip('/')
        self.api_key = api_key
        self.timeout = timeout
        self.max_retries = max_retries
        self.retry_delay = retry_delay
        self.token_storage = TokenStorage()
        
        # Create httpx client
        self._client = httpx.AsyncClient(
            timeout=httpx.Timeout(timeout),
            headers={
                "User-Agent": "Plinto-Python-SDK/1.0.0",
                "Content-Type": "application/json",
                "Accept": "application/json",
            },
        )
        
        # Rate limiting
        self._rate_limit_remaining: Optional[int] = None
        self._rate_limit_reset: Optional[float] = None
        self._rate_limit_lock = asyncio.Lock()
    
    async def __aenter__(self):
        return self
    
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.close()
    
    async def close(self):
        """Close the HTTP client"""
        await self._client.aclose()
    
    def set_tokens(
        self,
        access_token: str,
        refresh_token: str,
        expires_in: int,
    ) -> None:
        """Set authentication tokens"""
        self.token_storage.set_tokens(access_token, refresh_token, expires_in)
    
    def clear_tokens(self) -> None:
        """Clear authentication tokens"""
        self.token_storage.clear_tokens()
    
    def _get_auth_headers(self) -> Dict[str, str]:
        """Get authentication headers"""
        headers = {}
        
        # Try access token first
        access_token = self.token_storage.get_access_token()
        if access_token:
            headers["Authorization"] = f"Bearer {access_token}"
        elif self.api_key:
            headers["X-API-Key"] = self.api_key
        
        return headers
    
    async def _handle_rate_limit(self, response: Response) -> None:
        """Handle rate limiting"""
        # Update rate limit info from headers
        if "X-RateLimit-Remaining" in response.headers:
            self._rate_limit_remaining = int(response.headers["X-RateLimit-Remaining"])
        
        if "X-RateLimit-Reset" in response.headers:
            self._rate_limit_reset = float(response.headers["X-RateLimit-Reset"])
        
        # If rate limited, wait for reset
        if response.status_code == 429:
            retry_after = None
            if "Retry-After" in response.headers:
                retry_after = int(response.headers["Retry-After"])
            elif self._rate_limit_reset:
                retry_after = max(0, int(self._rate_limit_reset - time.time()))
            
            raise RateLimitError(
                "Rate limit exceeded",
                status_code=429,
                response_data=await self._safe_json(response),
                retry_after=retry_after,
            )
    
    async def _refresh_token(self) -> bool:
        """Attempt to refresh the access token"""
        refresh_token = self.token_storage.get_refresh_token()
        if not refresh_token:
            return False
        
        try:
            response = await self._client.post(
                f"{self.base_url}/api/v1/auth/refresh",
                json={"refresh_token": refresh_token},
                headers={"Content-Type": "application/json"},
            )
            
            if response.status_code == 200:
                data = response.json()
                self.token_storage.set_tokens(
                    data["access_token"],
                    data["refresh_token"],
                    data["expires_in"],
                )
                return True
        except Exception:
            pass
        
        # If refresh failed, clear tokens
        self.token_storage.clear_tokens()
        return False
    
    async def _safe_json(self, response: Response) -> Dict[str, Any]:
        """Safely parse JSON response"""
        try:
            return response.json()
        except (json.JSONDecodeError, ValueError):
            return {"message": response.text or "Unknown error"}
    
    def _handle_error_response(self, response: Response, data: Dict[str, Any]) -> None:
        """Handle error responses and raise appropriate exceptions"""
        status_code = response.status_code
        message = data.get("detail") or data.get("message") or f"HTTP {status_code}"
        
        if status_code == 400:
            # Check for validation errors
            if "errors" in data:
                raise ValidationError(
                    message,
                    status_code,
                    data,
                    field_errors=data.get("errors", {}),
                )
            raise ValidationError(message, status_code, data)
        elif status_code == 401:
            raise AuthenticationError(message, status_code, data)
        elif status_code == 403:
            raise PermissionError(message, status_code, data)
        elif status_code == 404:
            raise NotFoundError(message, status_code, data)
        elif status_code == 429:
            retry_after = None
            if "Retry-After" in response.headers:
                retry_after = int(response.headers["Retry-After"])
            raise RateLimitError(message, status_code, data, retry_after)
        elif 500 <= status_code < 600:
            raise ServerError(message, status_code, data)
        else:
            raise PlintoError(message, status_code, data)
    
    async def _make_request(
        self,
        method: str,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        files: Optional[Dict[str, Any]] = None,
    ) -> Response:
        """Make HTTP request with authentication and error handling"""
        url = urljoin(f"{self.base_url}/", path.lstrip('/'))
        
        # Prepare headers
        request_headers = {**self._get_auth_headers()}
        if headers:
            request_headers.update(headers)
        
        # Handle file uploads
        if files:
            # Remove content-type for multipart uploads
            request_headers.pop("Content-Type", None)
        
        for attempt in range(self.max_retries + 1):
            try:
                # Check rate limit before making request
                async with self._rate_limit_lock:
                    if (
                        self._rate_limit_remaining is not None
                        and self._rate_limit_remaining <= 0
                        and self._rate_limit_reset
                        and time.time() < self._rate_limit_reset
                    ):
                        wait_time = max(0, self._rate_limit_reset - time.time())
                        await asyncio.sleep(wait_time)
                
                # Make the request
                response = await self._client.request(
                    method,
                    url,
                    json=json_data,
                    params=params,
                    headers=request_headers,
                    files=files,
                )
                
                # Handle rate limiting
                await self._handle_rate_limit(response)
                
                # If unauthorized and we have a refresh token, try to refresh
                if response.status_code == 401 and self.token_storage.get_refresh_token():
                    if await self._refresh_token():
                        # Update headers with new token and retry
                        request_headers.update(self._get_auth_headers())
                        response = await self._client.request(
                            method,
                            url,
                            json=json_data,
                            params=params,
                            headers=request_headers,
                            files=files,
                        )
                
                # Handle errors
                if not response.is_success:
                    data = await self._safe_json(response)
                    self._handle_error_response(response, data)
                
                return response
                
            except (httpx.ConnectError, httpx.TimeoutException) as e:
                if attempt == self.max_retries:
                    raise NetworkError(f"Network error: {str(e)}", e)
                
                # Exponential backoff
                await asyncio.sleep(self.retry_delay * (2 ** attempt))
            
            except (RateLimitError, AuthenticationError, ValidationError) as e:
                # Don't retry these errors
                raise e
            
            except Exception as e:
                if attempt == self.max_retries:
                    raise NetworkError(f"Unexpected error: {str(e)}", e)
                
                await asyncio.sleep(self.retry_delay)
        
        raise NetworkError("Max retries exceeded")
    
    async def get(
        self,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Response:
        """Make GET request"""
        return await self._make_request("GET", path, params=params, headers=headers)
    
    async def post(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
        files: Optional[Dict[str, Any]] = None,
    ) -> Response:
        """Make POST request"""
        return await self._make_request(
            "POST", path, json_data=json_data, params=params, headers=headers, files=files
        )
    
    async def patch(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Response:
        """Make PATCH request"""
        return await self._make_request("PATCH", path, json_data=json_data, params=params, headers=headers)
    
    async def put(
        self,
        path: str,
        json_data: Optional[Dict[str, Any]] = None,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Response:
        """Make PUT request"""
        return await self._make_request("PUT", path, json_data=json_data, params=params, headers=headers)
    
    async def delete(
        self,
        path: str,
        params: Optional[Dict[str, Any]] = None,
        headers: Optional[Dict[str, str]] = None,
    ) -> Response:
        """Make DELETE request"""
        return await self._make_request("DELETE", path, params=params, headers=headers)

File: packages/typescript-sdk/src/http-client.ts
/**
 * HTTP client for the Janua TypeScript SDK
 */

import type {
  RequestConfig,
  HttpResponse,
  RateLimitInfo,
  JanuaConfig
} from './types';
import {
  JanuaError,
  NetworkError,
  RateLimitError,
  ConfigurationError
} from './errors';
import { TokenManager, RetryUtils, EventEmitter } from './utils';
import type { SdkEventMap } from './types';

/**
 * HTTP client with automatic token refresh and retry logic
 */
export class HttpClient extends EventEmitter<SdkEventMap> {
  private config: Required<Pick<JanuaConfig, 'baseURL' | 'timeout' | 'retryAttempts' | 'retryDelay'>>;
  private tokenManager: TokenManager;
  private refreshPromise: Promise<void> | null = null;

  constructor(config: JanuaConfig, tokenManager: TokenManager) {
    super();

    this.config = {
      baseURL: config.baseURL,
      timeout: config.timeout || 30000,
      retryAttempts: config.retryAttempts || 3,
      retryDelay: config.retryDelay || 1000
    };

    this.tokenManager = tokenManager;
  }

  /**
   * Make HTTP request with automatic token handling
   */
  async request<T = unknown>(config: RequestConfig): Promise<HttpResponse<T>> {
    return this.executeWithRetry(async () => {
      // Add authorization header if not skipped and token exists
      if (!config.skipAuth) {
        const accessToken = await this.tokenManager.getAccessToken();
        if (accessToken) {
          config.headers = {
            ...config.headers,
            Authorization: `Bearer ${accessToken}`
          };
        }
      }

      // Build full URL
      const url = this.buildUrl(config.url, config.params);

      // Make request
      const response = await this.makeRequest(url, config);

      const status = response.status ?? 200;
      const headers = response.headers ?? {};

      // Handle rate limiting
      if (status === 429) {
        const rateLimitInfo = this.parseRateLimitHeaders(headers);
        throw new RateLimitError('Rate limit exceeded', rateLimitInfo);
      }

      // Handle authentication errors
      if (status === 401 && !config.skipAuth) {
        await this.handleAuthError(config);
        // Retry the request with refreshed token
        return this.request(config);
      }

      // Handle API errors
      if (status >= 400) {
        await this.handleApiError(response);
      }

      return response as HttpResponse<T>;
    });
  }

  /**
   * Execute request with retry logic
   */
  private async executeWithRetry<T>(
    operation: () => Promise<T>
  ): Promise<T> {
    return RetryUtils.withRetry(operation, {
      maxAttempts: this.config.retryAttempts,
      initialDelay: this.config.retryDelay,
      backoffMultiplier: 2,
      maxDelay: 30000
    });
  }

  /**
   * Build full URL with query parameters
   */
  private buildUrl(path: string, params?: Record<string, any>): string {
    const url = new URL(path, this.config.baseURL);

    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          if (Array.isArray(value)) {
            value.forEach(v => url.searchParams.append(key, String(v)));
          } else {
            url.searchParams.set(key, String(value));
          }
        }
      });
    }

    return url.toString();
  }

  /**
   * Make the actual HTTP request
   */
  private async makeRequest(url: string, config: RequestConfig): Promise<HttpResponse> {
    const requestInit: RequestInit = {
      method: config.method,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': this.getUserAgent(),
        ...config.headers
      },
      signal: this.createAbortSignal(config.timeout || this.config.timeout)
    };

    if (config.data && config.method !== 'GET' && config.method !== 'DELETE') {
      requestInit.body = JSON.stringify(config.data);
    }

    try {
      const response = await fetch(url, requestInit);

      // Parse response
      let data: unknown;
      const contentType = response.headers.get('content-type');

      if (contentType && contentType.includes('application/json')) {
        const text = await response.text();
        data = text ? JSON.parse(text) : null;
      } else {
        data = await response.text();
      }

      return {
        data,
        status: response.status,
        statusText: response.statusText,
        headers: this.headersToObject(response.headers)
      };
    } catch (error) {
      if (error instanceof Error && error.name === 'AbortError') {
        throw new NetworkError('Request timeout');
      }
      throw new NetworkError('Network request failed', error instanceof Error ? error : undefined);
    }
  }

  /**
   * Handle authentication errors
   */
  private async handleAuthError(_originalConfig: RequestConfig): Promise<void> {
    // Prevent multiple simultaneous refresh attempts
    if (this.refreshPromise) {
      await this.refreshPromise;
      return;
    }

    try {
      this.refreshPromise = this.refreshTokens();
      await this.refreshPromise;
    } finally {
      this.refreshPromise = null;
    }
  }

  /**
   * Refresh access tokens
   */
  private async refreshTokens(): Promise<void> {
    const refreshToken = await this.tokenManager.getRefreshToken();

    if (!refreshToken) {
      this.emit('auth:signedOut', {});
      throw new JanuaError('No refresh token available', 'AUTHENTICATION_ERROR');
    }

    try {
      const response = await this.request<{
        access_token: string;
        refresh_token: string;
        expires_in: number;
      }>({
        method: 'POST',
        url: '/api/v1/auth/refresh',
        data: { refresh_token: refreshToken },
        skipAuth: true
      });

      // Store new tokens
      const expiresAt = Date.now() + (response.data.expires_in * 1000);
      await this.tokenManager.setTokens({
        access_token: response.data.access_token,
        refresh_token: response.data.refresh_token,
        expires_at: expiresAt
      });

      this.emit('token:refreshed', { tokens: { ...response.data, token_type: 'bearer' as const } });
    } catch (error) {
      // Clear tokens on refresh failure
      await this.tokenManager.clearTokens();
      this.emit('token:expired', {});
      this.emit('auth:signedOut', {});
      throw error;
    }
  }

  /**
   * Handle API errors
   */
  private async handleApiError(response: HttpResponse): Promise<never> {
    let errorData: { error?: string; message?: string; detail?: string; details?: Record<string, unknown> };

    try {
      errorData = typeof response.data === 'string'
        ? { message: response.data }
        : (response.data as typeof errorData) || { message: 'Unknown error occurred' };
    } catch {
      errorData = { message: 'Unknown error occurred' };
    }

    const apiError = {
      error: errorData.error || 'API_ERROR',
      message: errorData.message || errorData.detail || 'An error occurred',
      details: errorData.details || {},
      status_code: response.status ?? 500
    };

    throw JanuaError.fromApiError(apiError);
  }

  /**
   * Parse rate limit headers
   */
  private parseRateLimitHeaders(headers: Record<string, string>): RateLimitInfo {
    return {
      limit: parseInt(headers['x-ratelimit-limit'] || '0', 10),
      remaining: parseInt(headers['x-ratelimit-remaining'] || '0', 10),
      reset: parseInt(headers['x-ratelimit-reset'] || '0', 10),
      retry_after: parseInt(headers['retry-after'] || '0', 10) || undefined
    };
  }

  /**
   * Convert Headers object to plain object
   */
  private headersToObject(headers: Headers): Record<string, string> {
    const result: Record<string, string> = {};
    headers.forEach((value: string, key: string) => {
      result[key] = value;
    });
    return result;
  }

  /**
   * Create abort signal for timeout
   */
  private createAbortSignal(timeout: number): AbortSignal {
    const controller = new AbortController();
    setTimeout(() => controller.abort(), timeout);
    return controller.signal;
  }

  /**
   * Get user agent string
   */
  private getUserAgent(): string {
    const sdkVersion = '1.0.0'; // This should be dynamically set

    if (typeof window !== 'undefined') {
      return `janua-typescript-sdk/${sdkVersion} (Browser)`;
    } else if (typeof process !== 'undefined') {
      return `janua-typescript-sdk/${sdkVersion} (Node.js ${process.version})`;
    } else {
      return `janua-typescript-sdk/${sdkVersion}`;
    }
  }

  /**
   * Convenience methods for different HTTP verbs
   */
  async get<T = any>(url: string, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({
      method: 'GET',
      url,
      ...config
    });
  }

  async post<T = unknown>(url: string, data?: unknown, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({
      method: 'POST',
      url,
      data,
      ...config
    });
  }

  async put<T = unknown>(url: string, data?: unknown, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({
      method: 'PUT',
      url,
      data,
      ...config
    });
  }

  async patch<T = unknown>(url: string, data?: unknown, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({
      method: 'PATCH',
      url,
      data,
      ...config
    });
  }

  async delete<T = unknown>(url: string, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({
      method: 'DELETE',
      url,
      ...config
    });
  }
}

/** Axios instance type (minimal interface for dynamic import) */
interface AxiosInstance {
  interceptors: {
    request: { use: (onFulfilled: (config: AxiosRequestConfig) => Promise<AxiosRequestConfig>) => void };
    response: { use: (onFulfilled: (response: AxiosResponse) => AxiosResponse, onRejected: (error: AxiosError) => Promise<never>) => void };
  };
  get: <T>(url: string, config?: Partial<AxiosRequestConfig>) => Promise<AxiosResponse<T>>;
  post: <T>(url: string, data?: unknown, config?: Partial<AxiosRequestConfig>) => Promise<AxiosResponse<T>>;
  put: <T>(url: string, data?: unknown, config?: Partial<AxiosRequestConfig>) => Promise<AxiosResponse<T>>;
  patch: <T>(url: string, data?: unknown, config?: Partial<AxiosRequestConfig>) => Promise<AxiosResponse<T>>;
  delete: <T>(url: string, config?: Partial<AxiosRequestConfig>) => Promise<AxiosResponse<T>>;
  (config: AxiosRequestConfig): Promise<AxiosResponse>;
}

interface AxiosRequestConfig {
  url?: string;
  method?: string;
  baseURL?: string;
  headers?: Record<string, string>;
  params?: Record<string, unknown>;
  data?: unknown;
  timeout?: number;
  skipAuth?: boolean;
  _retry?: boolean;
}

interface AxiosResponse<T = unknown> {
  data: T;
  status: number;
  statusText: string;
  headers: Record<string, string>;
  config: AxiosRequestConfig;
}

interface AxiosError {
  config: AxiosRequestConfig;
  response?: AxiosResponse<{ error?: string; message?: string; detail?: string; details?: Record<string, unknown> }>;
  request?: unknown;
  message: string;
}

interface AxiosStatic {
  create: (config: { baseURL: string; timeout: number; headers: Record<string, string> }) => AxiosInstance;
}

/**
 * Axios adapter for environments that prefer axios
 */
export class AxiosHttpClient extends EventEmitter<SdkEventMap> {
  private axiosLib: AxiosStatic;
  private axios!: AxiosInstance;
  private tokenManager: TokenManager;
  private refreshPromise: Promise<void> | null = null;

  constructor(config: JanuaConfig, tokenManager: TokenManager) {
    super();

    this.tokenManager = tokenManager;

    try {
      // Try to import axios
      this.axiosLib = require('axios');
      this.setupAxiosInstance(config);
    } catch (error) {
      throw new ConfigurationError('Axios is not available. Please install axios or use the default fetch client.');
    }
  }

  private setupAxiosInstance(config: JanuaConfig): void {
    this.axios = this.axiosLib.create({
      baseURL: config.baseURL,
      timeout: config.timeout || 30000,
      headers: {
        'Content-Type': 'application/json'
      }
    });

    // Request interceptor for auth
    this.axios.interceptors.request.use(async (axiosConfig: AxiosRequestConfig) => {
      if (!axiosConfig.skipAuth) {
        const accessToken = await this.tokenManager.getAccessToken();
        if (accessToken) {
          axiosConfig.headers = axiosConfig.headers || {};
          axiosConfig.headers.Authorization = `Bearer ${accessToken}`;
        }
      }
      return axiosConfig;
    });

    // Response interceptor for error handling
    this.axios.interceptors.response.use(
      (response: AxiosResponse) => response,
      async (error: AxiosError): Promise<never> => {
        const originalRequest = error.config;

        if (error.response?.status === 401 && !originalRequest._retry && !originalRequest.skipAuth) {
          originalRequest._retry = true;

          try {
            await this.handleAuthError();
            return this.axios(originalRequest) as never;
          } catch (refreshError) {
            return Promise.reject(refreshError);
          }
        }

        // Convert axios error to our error format
        if (error.response) {
          const apiError = {
            error: error.response.data?.error || 'API_ERROR',
            message: error.response.data?.message || error.response.data?.detail || error.message,
            details: error.response.data?.details || {},
            status_code: error.response.status
          };
          throw JanuaError.fromApiError(apiError);
        } else if (error.request) {
          throw new NetworkError('Network request failed');
        } else {
          throw new NetworkError('Request setup failed');
        }
      }
    );
  }

  private async handleAuthError(): Promise<void> {
    if (this.refreshPromise) {
      await this.refreshPromise;
      return;
    }

    try {
      this.refreshPromise = this.refreshTokens();
      await this.refreshPromise;
    } finally {
      this.refreshPromise = null;
    }
  }

  private async refreshTokens(): Promise<void> {
    const refreshToken = await this.tokenManager.getRefreshToken();

    if (!refreshToken) {
      this.emit('auth:signedOut', {});
      throw new JanuaError('No refresh token available', 'AUTHENTICATION_ERROR');
    }

    try {
      const response = await this.axios.post<{ access_token: string; refresh_token: string; expires_in: number }>('/api/v1/auth/refresh', {
        refresh_token: refreshToken
      }, { skipAuth: true });

      const expiresAt = Date.now() + (response.data.expires_in * 1000);
      await this.tokenManager.setTokens({
        access_token: response.data.access_token,
        refresh_token: response.data.refresh_token,
        expires_at: expiresAt
      });

      this.emit('token:refreshed', { tokens: { ...response.data, token_type: 'bearer' as const } });
    } catch (error) {
      await this.tokenManager.clearTokens();
      this.emit('token:expired', {});
      this.emit('auth:signedOut', {});
      throw error;
    }
  }

  async request<T = unknown>(config: RequestConfig): Promise<HttpResponse<T>> {
    try {
      const response = await this.axios({
        method: config.method,
        url: config.url,
        data: config.data,
        params: config.params as Record<string, unknown>,
        headers: config.headers,
        timeout: config.timeout,
        skipAuth: config.skipAuth
      });

      return {
        data: response.data as T,
        status: response.status,
        statusText: response.statusText,
        headers: response.headers
      };
    } catch (error) {
      throw error; // Already converted by interceptor
    }
  }

  // Convenience methods
  async get<T = unknown>(url: string, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({ method: 'GET', url, ...config });
  }

  async post<T = unknown>(url: string, data?: unknown, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({ method: 'POST', url, data, ...config });
  }

  async put<T = unknown>(url: string, data?: unknown, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({ method: 'PUT', url, data, ...config });
  }

  async patch<T = unknown>(url: string, data?: unknown, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({ method: 'PATCH', url, data, ...config });
  }

  async delete<T = unknown>(url: string, config?: Partial<RequestConfig>): Promise<HttpResponse<T>> {
    return this.request<T>({ method: 'DELETE', url, ...config });
  }
}

/**
 * Factory function to create appropriate HTTP client
 */
export function createHttpClient(config: JanuaConfig, tokenManager: TokenManager): HttpClient | AxiosHttpClient {
  // Check if axios is available and preferred
  if (config.environment && config.environment !== 'browser' as any) {
    try {
      require.resolve('axios');
      return new AxiosHttpClient(config, tokenManager);
    } catch {
      // Axios not available, use fetch client
    }
  }

  return new HttpClient(config, tokenManager);
}
